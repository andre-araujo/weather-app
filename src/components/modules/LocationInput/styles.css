.container {
  width: 100vw;
  height: 100vh;
  overflow: auto;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  padding: 50px;
  background: rgba(255, 255, 255, 0.9);
  position: fixed;
  z-index: 10;
  top: 0;
  left: 0;
}

.input {
  font-size: 40px;
  outline: none;
  border: none;
  border-bottom: 5px solid #333;
  background: none;
  transition: .2s border-color;
  text-align: center;
  width: 100%;
  max-width: 400px;
}

.input:focus {
  border-color: orange;
}

.icon {
  margin-right: 10px;
}

.row {
  margin-bottom: 15px;
}
