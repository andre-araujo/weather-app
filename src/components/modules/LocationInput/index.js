import React, { Component } from 'react';
import { func, string } from 'prop-types';

import styles from './styles.less';
import Icon from '../../elements/Icon';
import Button from '../../elements/Button';
import getLatLong from '../../../lib/getLatLong';

class LocationInput extends Component {
  state = {
    location: '',
  }

  handleInput = (e) => {
    this.setState({
      location: e.target.value,
    });
  }

  handleGetCoords = () => {
    const { onGetCoords } = this.props;

    getLatLong(({ coords }) => {
      onGetCoords({
        latitude: coords.latitude,
        longitude: coords.longitude,
      });
    }, (error) => {
      onGetCoords({
        error,
      });
    });
  }

  handleSearch = (e) => {
    const { onSearch } = this.props;
    e.preventDefault();

    onSearch(this.state.location);
  }

  render() {
    const { location } = this.state;

    return (
      <form className={styles.container} onSubmit={this.handleSearch}>
        <div className={styles.row}>
          <input className={styles.input} type="text" onChange={this.handleInput} value={location} placeholder="Ex: Rio de Janeiro" />
        </div>
        {this.props.error && (
          <div className={styles.row}>
            <p className={styles.error}>{this.props.error}</p>
          </div>
        )}
        <div className={styles.row}>
          <Button type="button" onClick={this.handleGetCoords} data-test="LocationInput_GetCoords">
            <span className={styles.icon}><Icon.Compass /></span>
            get location
          </Button>
        </div>
        <div className={styles.row}>
          <Button type="submit">Search</Button>
        </div>
      </form>
    );
  }
}

LocationInput.defaultProps = {
  error: null,
};

LocationInput.propTypes = {
  error: string,
  onSearch: func.isRequired,
  onGetCoords: func.isRequired,
};

export default LocationInput;
