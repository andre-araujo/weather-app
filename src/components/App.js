import React, { Component, Fragment } from 'react';

import { getImage } from '../services/imageAPI';
import { getWeatherByLocationName, getWeatherByLatLong } from '../services/yahooAPI';

import imageFactory from '../services/factories/imageFactory';
import weatherForecastFactory from '../services/factories/weatherForecastFactory';

import WeatherForecast from './modules/WeatherForecast';
import ImageBackground from './modules/ImageBackground';
import LocationInput from './modules/LocationInput';

class App extends Component {
  state={
    imageURL: '#',
    weather: null,
  }

  componentDidMount() {
    getImage()
      .then(resp => resp.json())
      .then(imageFactory)
      .then(({ imageURL }) => this.setState({ imageURL }));
  }

  getWeatherByLocationName = (locationName) => {
    if (locationName.length === 0) {
      this.setState({ error: 'Empty location' });
      return;
    }

    getWeatherByLocationName(locationName)
      .then(resp => resp.json())
      .then((resp) => {
        if (!resp.query || !resp.query.results) { throw new Error('Invalid location'); }
        return resp;
      })
      .then(weatherForecastFactory)
      .then(weather => this.setState({ weather, error: null }))
      .catch(fetchError => this.setState({ error: fetchError.message }));
  }

  getWeatherByLatLong = ({ latitude, longitude, error }) => {
    if (error) {
      this.setState({ error: error.message });
      return;
    }

    getWeatherByLatLong(latitude, longitude)
      .then(resp => resp.json())
      .then(weatherForecastFactory)
      .then(weather => this.setState({ weather, error: null }))
      .catch(fetchError => this.setState({ error: fetchError.message }));
  }

  render() {
    const { imageURL, weather, error } = this.state;

    return (
      <Fragment>
        <ImageBackground imageURL={imageURL}>
          {
            weather &&
            <WeatherForecast weather={weather} />
          }
        </ImageBackground>
        {
          !weather &&
          <LocationInput
            error={error}
            onGetCoords={this.getWeatherByLatLong}
            onSearch={this.getWeatherByLocationName}
          />
        }
      </Fragment>
    );
  }
}

export default App;
