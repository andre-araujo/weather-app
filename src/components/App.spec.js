import React from 'react';
import { shallow } from 'enzyme';

import App from './App';
import LocationInput from './modules/LocationInput';

jest.mock('../services/factories/imageFactory', val => val);
jest.mock('../services/factories/weatherForecastFactory', val => val);

jest.mock('../services/imageAPI');
jest.mock('../services/yahooAPI');

const { getImage } = require('../services/imageAPI');
const { getWeatherByLocationName, getWeatherByLatLong } = require('../services/yahooAPI');

const fetchMock = () => Promise.resolve({
  json: () => ({}),
});

const getImageImp = jest.fn(fetchMock);
const getWeatherByLocationNameImp = jest.fn(fetchMock);
const getWeatherByLatLongImp = jest.fn(fetchMock);

beforeEach(() => {
  getImageImp.mockClear();
  getWeatherByLocationNameImp.mockClear();
  getWeatherByLatLongImp.mockClear();

  getImage.mockImplementation(getImageImp);
  getWeatherByLocationName.mockImplementation(getWeatherByLocationNameImp);
  getWeatherByLatLong.mockImplementation(getWeatherByLatLongImp);
});

describe('App component', () => {
  it('should match snapshot', () => {
    expect(shallow(<App />)).toMatchSnapshot();
  });

  it('should call getImage when componentDidMount', () => {
    shallow(<App />);

    expect(getImageImp).toHaveBeenCalledTimes(1);
  });

  it('should call setState with error message when getImage is rejected', (done) => {
    getImage.mockImplementation(() => Promise.reject(Error('getImage error')));

    const app = shallow(<App />);

    setImmediate(() => {
      expect(app.state('error')).toEqual('getImage error');
      done();
    });
  });

  describe('getWeatherByLocationName', () => {
    it('should set error state to Empty location when LocationInput\'s onSearch is called with no location', (done) => {
      const app = shallow(<App />);

      app.find(LocationInput).props().onSearch();

      setImmediate(() => {
        expect(app.state('error')).toEqual('Empty location');
        done();
      });
    });

    it('should call getWeatherByLocationName service when getWeatherByLocationName is called with string length > 1', (done) => {
      const app = shallow(<App />);

      app.find(LocationInput).props().onSearch('Rio de Janeiro');

      setImmediate(() => {
        expect(app.state('error')).toEqual('Invalid location');
        expect(getWeatherByLocationNameImp).toHaveBeenCalledWith('Rio de Janeiro');
        done();
      });
    });

    it('should set error state to Invalid location when getWeatherByLocationName service returns no results', (done) => {
      const app = shallow(<App />);

      getWeatherByLocationNameImp.mockImplementation(() => Promise.resolve({
        json: () => ({
          query: {
            results: null,
          },
        }),
      }));

      app.find(LocationInput).props().onSearch('Rio de Janeiro');

      setImmediate(() => {
        expect(app.state('error')).toEqual('Invalid location');
        done();
      });
    });

    it('should set weather state with getWeatherByLocationName service response', (done) => {
      const app = shallow(<App />);

      getWeatherByLocationNameImp.mockImplementation(() => Promise.resolve({
        json: () => ({
          query: {
            results: [
              {
                data: [],
              },
            ],
          },
        }),
      }));

      app.find(LocationInput).props().onSearch('Rio de Janeiro');

      setImmediate(() => {
        expect(app.state('weather')).toEqual({ query: { results: [{ data: [] }] } });
        expect(app).toMatchSnapshot();
        done();
      });
    });
  });
});
