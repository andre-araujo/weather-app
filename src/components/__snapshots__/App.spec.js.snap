// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component getWeatherByLocationName should set weather state with getWeatherByLocationName service response 1`] = `
<React.Fragment>
  <ImageBackground
    imageURL="#"
  />
  <LocationInput
    error={null}
    onGetCoords={[Function]}
    onSearch={[Function]}
  />
</React.Fragment>
`;

exports[`App component should match snapshot 1`] = `
<React.Fragment>
  <ImageBackground
    imageURL="#"
  />
  <LocationInput
    error={null}
    onGetCoords={[Function]}
    onSearch={[Function]}
  />
</React.Fragment>
`;
